engine/crates/graphql-schema-diff/src/traverse_schemas.rs

```diff
@@ -0,0 +1,216 @@
+use crate::{ast, ConstValue, DefinitionKind, DiffMap, DiffState, Positioned};
+use std::{collections::hash_map::Entry, hash::Hash};
+
+pub(crate) fn traverse_schemas<'a>([source, target]: [&'a ast::ServiceDocument; 2], state: &mut DiffState<'a>) {
+    let schema_size_approx = source.definitions.len().max(target.definitions.len());
+    state.types_map.reserve(schema_size_approx);
+    state.fields_map.reserve(schema_size_approx);
+
+    traverse_source(source, state);
+    traverse_target(target, state);
+}
+
+fn traverse_source<'a>(source: &'a ast::ServiceDocument, state: &mut DiffState<'a>) {
```

Comment:
Without knowing too much about the idea behind this, why do we need two quite similar functions for source and target? Could they be combined into a single function?

Reply:
They have similar loops with different content — we can combine by introducing more structure (visitor pattern like). I'll give a shot at it and see if that ends up simplifying or making things more obscure.

Reply:
I've given it a try and introducing a visitor here for two instances that visit the same things feels a bit overengineered in practice.